### ISSUE TITLE
Bug hunt: spline returns 0 at the right boundary knot, CV :observation shifts the ODE start when t0=nothing, soft-tree degenerate guard misses multi-output trees

### ISSUE THREAD

**Issue by manuhuth** [manuhuth/NoLimits.jl#304]
Orchestrated bug hunt over cross-validation and the universal function approximators (splines, neural networks, soft trees, normalizing flows), run against `origin/main` at `03005aba4` (v0.2.7 plus the merged #296-#303 fixes). 11 slices, each with a repro script and an independent look at the implicated source. Three defects survived verification; two of them were re-confirmed independently of the agent that found them.

## 1. `bspline_eval` / `bspline_basis` silently return `0.0` at `x == knots[end]` (high)

`src/utils/Splines.jl:62`:

```julia
mu = x == knots[end] ? m - 1 : searchsortedlast(knots, x)
```

At the right boundary the knot span index is set to `m - 1` (where `m = length(knots)`), but the last existing basis function lives at index `n = length(knots) - degree - 1`. Every index the Cox-de Boor triangle then derives falls outside the valid window, or lands on a degenerate span whose knot differences are zero, so **every basis function evaluates to exactly `0.0`** and `bspline_eval` returns `0.0` regardless of the coefficients. One step to the left the basis is perfectly correct.

`_bspline_validate` (`Splines.jl:34-44`) accepts `x == knots[end]` as in range (strict `<` / `>` comparisons only), and both docstrings state the domain is the closed interval `[knots[1], knots[end]]`, so this is a wrong value inside the documented domain, not an extrapolation guard.

```julia
using NoLimits
knots = [0.0, 0, 0, 0, 0.3, 0.6, 1, 1, 1, 1]; c = [1.0, 2, 3, 4, 5, 6]
NoLimits.bspline_eval(0.999999, c, knots, 3)   # 5.999992500008037
NoLimits.bspline_eval(1.0,      c, knots, 3)   # 0.0        <-- expected 6.0
sum(NoLimits.bspline_basis(1.0, knots, 3))     # 0.0        <-- expected 1.0

k2 = collect(0.0:0.2:1.0)                       # uniform knots, degree 1
NoLimits.bspline_eval(1.0, ones(length(k2) - 2), k2, 1)  # 0.0  <-- expected 1.0
```

It is not specific to clamped knot vectors: any knot vector evaluated exactly at `knots[end]` returns zero. Both knot vectors that appear in the docs end at `1.0` (`[0.0, 0.0, 0.0, 0.5, 1.0, 1.0, 1.0]` in `docs/src/model-building/fixed-effects.md:130`, `collect(range(0.0, 1.0; length=6))` in `docs/src/model-building/pre-differential-equation.md:99`), so any model whose spline input is a covariate or a time normalized to `[0, 1]` produces a silent `0.0` from the spline term at every row where the input is exactly `1.0`. `bspline_eval` is what the generated model function calls for every `SplineParameters` block (`src/model/FixedEffects.jl:665,667`), so the corruption reaches the outcome model, the gradient and the fit. The returned value is a clean float rather than `NaN`, so nothing flags it.

`test/splines_tests.jl` only evaluates at `x = 0.25`, which is why this was never caught.

Suggested fix, one line:

```julia
mu = min(searchsortedlast(knots, x), m - degree - 1)
```

Verified against the real implementation with only that line changed: bit-identical on the whole interior (max absolute difference exactly `0.0` over a 2000-point grid on `[0, 1)`), returns `6.0` at `x = 1.0` for the example above, and partition of unity then holds to `4.4e-16` across the closed interval instead of failing by `1.0` at the endpoint. A regression test evaluating at `knots[end]` and asserting `sum(bspline_basis) ≈ 1` would cover it.

Separate and secondary: for a non-clamped knot vector, `_bspline_validate` accepts the whole of `[knots[1], knots[end]]` even though the spline's actual domain is `[knots[degree+1], knots[n+1]]`. Worth a decision, but distinct from the bug above.

## 2. `fit_cv(kind = :observation)` on an ODE model built with `t0 = nothing` anchors the initial condition at different times in the training and the scoring model (high)

When `t0 === nothing`, each individual's integration start is `minimum(tvals)` over **the rows present in the DataModel that was passed in** (`src/data_model/DataModel.jl:1630-1635`), and `@initialDE` states are injected at `tspan[1]`. `_rebuild_dm` (`src/estimation/cv.jl:102-112`) correctly forwards `cfg.t0`, so a fold's DataModel recomputes its own time origin from whichever rows survived the split.

Under `kind = :observation` (`cv.jl:194-224`) each individual's observation rows are round-robined across folds. For the fold that receives an individual's temporally first observation as its **test** row, `dm_train` no longer contains that row, so the start moves forward. The paired `dm_test` is built from `train ∪ test` (`cv.jl:772-779`), which restores all of that individual's rows and therefore the original earlier start. Within one fold the model that is fitted and the model that scores it disagree about when the trajectory begins.

Confirmed through a real `cross_validate` call (4 individuals at t = 0, 1, 2, 3; `D(x1) ~ -k*x1`, `@initialDE x1 = 10.0`, `t0 = nothing`, `kind = :observation`, k = 4). Every fold contains at least one such individual:

```
fold 2
  train tspans = [(0.0, 3.0), (1.0, 3.0), (1.0, 3.0), (0.0, 2.0)]
  test  tspans = [(0.0, 3.0), (0.0, 3.0), (0.0, 3.0), (0.0, 3.0)]
```

For those individuals the `x1 = 10` initial condition fires at `t = 1` during fitting and at `t = 0` during scoring, shifting the whole predicted curve by one time unit (`y` at t = 1, 2, 3, 4 becomes `[10.0, 8.187, 6.703, 5.488]` instead of `[8.187, 6.703, 5.488, 4.493]`). The fold's parameters and EBEs are calibrated against a curve whose initial condition fires at the wrong physical time, then scored against one where it does not. The result is a split-dependent bias in the reported predictive log-likelihood, with no error and no warning.

Scope: `t0 = nothing` combined with `kind = :observation`, for individuals whose earliest row is a plain observation. Dose-anchored models are unaffected, because event rows are retained in every fold (`cv.jl:207-218`), and the default `t0 = 0.0` is unaffected, because `start = min(t0, lo)` stays pinned at `0.0`. Nothing in `cv.jl`'s docstrings or `docs/src/estimation/cv.md` mentions a `t0` interaction, and no CV test uses an ODE model, so nothing asserts this on purpose.

The natural fix is for `_rebuild_dm` to pass the parent DataModel's already-resolved per-individual start times instead of letting each fold re-derive them, so that every fold, train and test alike, inherits the full-data time origin. Refusing `t0 = nothing` together with `kind = :observation` would at least make it loud rather than silent.

## 3. The soft-tree degenerate-start guard misses multi-output trees whose leaves are constant within each output row (medium)

`_warn_degenerate_soft_trees` (`src/estimation/common.jl:3363-3376`) flattens the whole leaf block, all `n_output * 2^depth` entries, and tests `all(isequal(first(leaves)), leaves)`. The zero-gradient condition is per output row: since `∂output_o/∂split = v[o, :] ⋅ ∂P/∂split` and `Σ_leaf ∂P[leaf]/∂split = ∂(Σ P)/∂split = 0`, the gradient with respect to every split parameter is exactly zero as soon as the leaves are constant **within each output row**, whether or not different rows share the same constant.

So a 2-output depth-2 tree started with row 1 all `0.3` and row 2 all `0.7` sits at exactly the saddle the guard exists to catch (ForwardDiff confirms `max |grad|` over the split parameters is `0.0`), the optimizer cannot move the splits, it reports success having trained a constant, and no warning is emitted. The equivalent single-output start does warn, as `test/softtrees_tests.jl:80-92` asserts.

`n_output` is a documented, supported keyword (`src/model/Parameters.jl:721-755`) with no caveat, and every existing soft-tree test uses `n_output = 1`, so the multi-output path is simply untested here. Fix: reshape the leaf block to `(n_output, 2^depth)` and warn when any row is constant, rather than comparing the flattened block against a single value.

## Verified clean

Recorded so these are not re-hunted soon. Each was exercised with repro scripts, not only read.

Cross-validation:

- Fold and partition construction for both `kind = :id` and `kind = :observation`: exhaustive, pairwise disjoint, floor/ceil balanced, across n and k in `{(6,2), (6,3), (7,3), (7,7), (5,5), (5,4)}`. Grouping is by id **value** (`_group_indices` builds a `Dict{Any, Vector{Int}}`), so scrambled, interleaved, non-contiguous and `String` ids all group correctly and no id is ever split between train and test under `:id`. The `n >= n_folds` and `n_folds >= 2` guards give clear errors. Same seed gives byte-identical splits.
- No training-fold leakage: `fit_model` only ever receives `dm_train`, built from disjoint rows. `:ebe` takes its EBEs from the training fit's cached EB modes, `:conditional` samples from a posterior built on `dm_train` only. No cross-fold state carryover: per-fold RNGs are seeded by index, so results are independent of fold order for both the serial and threaded branches.
- `:ebe` matches the closed-form Gaussian posterior-mode plug-in to `1.7e-9` across all scored rows.
- Unseen-individual modes under `kind = :id`: `:mean` matches `logpdf(Normal(a, σ), y)` to `1.8e-15` for every held-out observation, with the fold aggregate matching an independent public-API refit to `7e-15`. `:montecarlo` converges on the true marginal `logpdf(Normal(a, sqrt(σ² + ω²)), y)` with RMSE `0.0273` at S = 2000 and `0.00865` at S = 20000, a `3.17x` reduction against the expected `sqrt(10) ≈ 3.16`, so the `logsumexp - log(n)` aggregation is unbiased. For a `LogNormal` random effect, `_re_prior_mean_or_zero` correctly supplies `exp(ω²/2)` and agrees with `predict(..., re_mode = :population)` to 6 significant figures.
- Aggregation arithmetic: `mean_test_loglikelihood` is the unweighted mean of per-fold totals as documented, `std_test_loglikelihood` and `mean_obs_loglikelihood` recompute by hand exactly, including with unequal fold sizes. `-Inf` from a genuinely ruled-out observation propagates honestly rather than being dropped, `NaN` from an evaluation failure is dropped but always with a warning naming the exact count (`cv.jl:822-828`). Per-observation rows map back to the correct original id under shuffled, non-contiguous and `String` ids. Repeated seeded calls are bit-identical.
- Keyword propagation into every fold fit, verified by observable effect for Laplace: `constants` (natural scale), `penalty`, `extra_objective`, `ode_args` / `ode_kwargs`. None of the prior, penalty or `extra_objective` terms leak into the held-out score, which is computed purely from the outcome distribution. An inner fit that throws propagates rather than being swallowed.
- DataModel integrity under subsetting: covariate columns, censoring flags and multiple-outcome assignments are re-parsed from the subset's own rows, so no row-versus-subject index mix-up is possible. Dose and event rows are retained in every fold. Nested and crossed random-effect levels resolve per effect, so a test individual sharing one level with training but not another is handled.

Universal function approximators:

- Estimator gradient correctness with UFAs in the path, which the previous round only covered for plain parametric models: `objective_and_gradient` versus `central_fdm(5, 1)`, componentwise, at 3+ parameter points, for MLE, MAP, Laplace, FOCEI and GHQuadrature, over models containing FFNN (Lux), soft-tree, spline and `NormalizingPlanarFlow` blocks. Relative errors `1e-12` to `1e-14`, with no zero or constant-factor-wrong components in any UFA block. Also clean: a random effect used as an NN input (`1e-6` to `1e-9`), the Laplace inner EB gradient and Hessian, and `use_trace_logdet_grad` true versus false agreeing to `1e-14`.
- Soft trees: internal nodes form a genuine nested binary partition of the leaves for depths 1 to 5, leaf probabilities are non-negative and sum to 1, gradients with respect to split weights, biases, leaf values and the input all match FD, second derivatives and depth-2 nested Duals are finite and match to `1e-11`, both temperature limits behave (mean of leaves as steepness goes to 0, exact hard-selected leaf as it goes to `1e12`), `|x|` up to `1e10` stays finite, `Float32` / `BigFloat` / `Dual` preserve eltype, and the flat parameter length matches `n_internal*input_dim + n_internal + n_output*2^depth` with a lossless destructure round trip.
- Neural networks: forward pass matches a hand-written `W*x .+ b` loop to `1.1e-16` over 6 architectures and all 8 registered activations plus softmax, so bias placement, output-activation handling and layer ordering are right and `_ffnn_nparams` matches what the forward pass reads. Lux glue reproduces `Lux.apply` through the reconstructor and the documented seeded init exactly. SimpleChains agrees with a hand-coded reference at every nested-Dual depth 0 to 6, i.e. both the native path and the `SC_MAX_DUAL_DEPTH` fallback are correct and agree at the boundary. Gradients with respect to parameters and inputs match FD for every backend.
- Splines away from the right endpoint: partition of unity, non-negativity and local support for degrees 1 to 3, exact polynomial reproduction to `1e-12`, continuity across interior knots including a repeated knot, coefficient gradients versus FD, `ForwardDiff` derivative in `x` to `1e-12`, and `Float32` consistency.
- UFA lifecycle: `get_params` returns the same coordinate count for `:untransformed` and `:transformed`, flat names are unique with nothing dropped or duplicated, `summarize` reports all 29 coordinates of a combined NN + spline + soft-tree model, Wald UQ returns matching counts on both scales, `get_residuals`, all four `predict` re_modes, `compute_shrinkage` and `show` all work. Save and load reproduces `get_params` and predictions bit-identically once the model is passed back to `load_fit`. `plot_fits` renders.
- `fit_cv` on a combined NN + spline + soft-tree model under both split kinds: every fold finite, every observation scored. Fold starting values cannot leak, because the parameter blocks are immutable and every fold refits from the same seeded `value`. No library-internal data-derived UFA hyperparameter exists (spline knots are user-supplied and static, and neither FFNN nor soft trees standardize their inputs), so the full-data-statistics leakage pattern does not arise inside `cross_validate`.

## Not covered

- Only Laplace was driven end-to-end through `fit_cv`. FOCEI, GHQuadrature, MLE, MAP, Pooled, SAEM and MCEM were reviewed at the source level, where the same `kwargs...` forwarding and the same `_cv_evaluate_*` dispatch apply uniformly, but were not run.
- Enzyme reverse mode was not compared against ForwardDiff for the UFA paths.
- `:conditional` seen-mode showed roughly `0.015` nat scatter against the analytic marginal at 20000 MC samples. The `:montecarlo` sample-scaling check above shows this estimator family converging at the expected rate, so this reads as ordinary Monte Carlo noise and is not reported as a defect.
